matchers/have_value.go

```diff
@@ -28,51 +31,44 @@ func HaveValue(matcher types.GomegaMatcher) types.GomegaMatcher {
 }
 
 type HaveValueMatcher struct {
-	Matcher types.GomegaMatcher // the matcher to apply to the "resolved" actual value.
-	failure string              // failure message, if any.
+	Matcher        types.GomegaMatcher // the matcher to apply to the "resolved" actual value.
+	resolvedActual interface{}         // the ("resolved") value.
 }
 
 func (m *HaveValueMatcher) Match(actual interface{}) (bool, error) {
 	val := reflect.ValueOf(actual)
-	for allowedIndirs := 32; allowedIndirs > 0; allowedIndirs-- {
+	for allowedIndirs := maxIndirections; allowedIndirs > 0; allowedIndirs-- {
 		// return an error if value isn't valid. Please note that we cannot
 		// check for nil here, as we might not deal with a pointer or interface
 		// at this point.
 		if !val.IsValid() {
-			m.failure = format.Message(
-				actual, "not to be <nil>")
-			return false, nil
+			return false, errors.New(format.Message(
+				actual, "not to be <nil>"))
 		}
 		switch val.Kind() {
 		case reflect.Ptr, reflect.Interface:
 			// resolve pointers and interfaces to their values, then rinse and
 			// repeat.
 			if val.IsNil() {
-				m.failure = format.Message(
-					actual, "not to be <nil>")
-				return false, nil
+				return false, errors.New(format.Message(
+					actual, "not to be <nil>"))
 			}
 			val = val.Elem()
 			continue
 		default:
 			// forward the final value to the specified matcher.
-			elem := val.Interface()
-			match, err := m.Matcher.Match(elem)
-			if !match {
-				m.failure = m.Matcher.FailureMessage(elem)
-			}
-			return match, err
+			m.resolvedActual = val.Interface()
+			return m.Matcher.Match(m.resolvedActual)
 		}
 	}
 	// too many indirections: extreme star gazing, indeed...?
-	m.failure = format.Message(actual, "indirecting too many times")
-	return false, nil
+	return false, errors.New(format.Message(actual, "too many indirections"))
 }
 
 func (m *HaveValueMatcher) FailureMessage(_ interface{}) (message string) {
-	return m.failure
+	return m.Matcher.NegatedFailureMessage(m.resolvedActual)
```

Comment:
all looks great now!  just this one line - you want `m.Matcher.FailureMessage` not `m.Matcher.NegatedFailureMessage` (sorry if I somehow implied that earlier - if so I miscommunicated)

Reply:
once this is tweaked I can pull this in - all else LGTM!

Reply:
That is something I obviously botches, so many thanks for spotting it. I'll have to think about a test to properly catch these types of mistakes.

Reply:
fixed.